src/tactic/rewrite_search/search.lean

```diff
@@ -0,0 +1,212 @@
+two vertices in a graph, and uses rewrite rules to find a path that connects the two sides.
+-/
+
+open tactic
+
+namespace tactic.rewrite_search
+
+/--
+An edge represents a proof that can get from one expression to another.
+It represents the fact that, starting from the vertex `fr`, the expression in `proof`
+can prove the vertex `to`.
+`how` contains information that the explainer will use to generate Lean code for the
+proof.
+-/
+meta structure edge :=
+(fr to : ℕ)
+(proof : tactic expr)
+(how   : how)
+
+/-- Converting an edge to a human-readable string. -/
+meta def edge.to_string : edge → format
+| e := format!"{e.fr} → {e.to}"
+
+meta instance edge.has_to_format : has_to_format edge := ⟨edge.to_string⟩
+
+/--
+A vertex represents an expression that is equivalent to either the left or right side
+of our initial equation.
+`id` is a numerical id used to refer to this vertex in the context of a single graph.
+`exp` is the expression this vertex represents.
+`pp` is the string format of the expression; we store this in the vertex to avoid
+recalculating it.
+`side` is whether this vertex descends from the left or right side of the equation.
+`parent` is the edge that originally added this vertex to the graph.
+-/
+meta structure vertex :=
+(id     : ℕ)
+(exp    : expr)
+(pp     : string)
+(side   : side)
+(parent : option edge)
+
+/--
+The graph represents two trees, one descending from each of the left and right sides
+of our initial equation.
+`conf` and `rules` determine what rewrites are used to generate new graph vertices.
+Here, the format of a rewrite rule is an expression for rewriting, plus a flag for the
+direction to apply it in.
+`vertices` maps vertex.id to vertex.
+`vmap` maps vertex.pp to vertex.id, so we can quickly find collisions.
+`solving_edge` represents a solution that will prove our target equation.
+It is an edge that would connect the two trees, so `solving_edge.fr` and `solving_edge.to`
+are vertices in different trees.
+-/
+meta structure graph :=
+(conf         : config)
+(rules        : list (expr × bool))
+(vertices     : buffer vertex)
+(vmap         : native.rb_map string ℕ)
+(solving_edge : option edge)
+
+/--
+Construct a graph to search for a proof of a given equation.
+
+This graph initially contains only two disconnected vertices corresponding to the two
+sides of the equation. When `find_proof` is called, we will run a search and add
+new vertices and edges.
+-/
+meta def mk_graph (conf : config) (rules : list (expr × bool)) (eq : expr) : tactic graph :=
+do (lhs, rhs) ← split_equation eq,
+  lhs_pp ← to_string <$> tactic.pp lhs,
+  rhs_pp ← to_string <$> tactic.pp rhs,
+  let lhs_vertex : vertex := ⟨0, lhs, lhs_pp, side.L, none⟩,
+  let rhs_vertex : vertex := ⟨1, rhs, rhs_pp, side.R, none⟩,
+  return ⟨conf, rules, [lhs_vertex, rhs_vertex].to_buffer,
+          native.rb_map.of_list [(lhs_pp, 0), (rhs_pp, 1)], none⟩
+
+variables (g : graph)
```

Comment:
I very much sense the presence of a state monad here. I won't insist on a refactor, but I suspect this would look really clean in that style.

Reply:
I don't understand this comment. I'm not saying it's wrong, I just don't understand. But I'm excited about the "won't insist on a refactor" part of your comment ;-)